hal/mbed_sleep_manager.c

```diff
@@ -196,7 +203,11 @@ void sleep_manager_sleep_auto(void)
 #ifdef MBED_DEBUG
     hal_sleep();
 #else
-    if (sleep_manager_can_deep_sleep()) {
+    if (sleep_manager_can_deep_sleep()
+#if DEVICE_LPTICKER && (LPTICKER_DELAY_TICKS > 0)
```

Comment:
If lp ticker's Timeout object is in use (i.e. `lp_ticker_get_timeout_pending()` returns `true`) and deep-sleep is not locked by other features (i.e. `lock_count==1`), then:
- `sleep_manager_can_deep_sleep()` returns `true` - indicates that we can put board into deep-sleep mode.
- `sleep_manager_sleep_auto()` puts the board into the sleep mode (not deep-sleep).

It looks like a design issue since we can not relay on `sleep_manager_can_deep_sleep()` function. 
From the user perspective it can be very misleading. From the testing perspective  `sleep_manager_can_deep_sleep()` function is used in tests to check if current environment state allows to go into deep-sleep mode which is to be tested. So in some cases the test may test sleep mode instead of required deep-sleep mode.

Reply:
> It looks like a design issue since we can not relay on sleep_manager_can_deep_sleep() function.

where this comes from?

Reply:
I agree with @mprse, `sleep_manager_can_deep_sleep()` shouldn't be updated this way. I do understand the purpose, but think this change should be test specific, not general.

Reply:
Please take a look at these changes https://github.com/ARMmbed/mbed-os/pull/7524/commits/c85856b813c3a596386cba7a54deb3fed5774069.

Reply:
Some possible alternatives I can think of are:
- Modify tests to suspend the OS before running the tests - this may be intrusive as you can no longer use most OS primitives
- Assert in tests using a different function - something like sleep_manager_can_deep_sleep_test()
- Assert something different in the tests, such as time actually spent in deep sleep

Do you have any thoughts or preferences on this @mprse and @fkjagodzinski?

Reply:
In my opinion first bullet is ok, but only for low level tests (HAL) where target specific drivers are tested, where the OS can be omitted. For driver layer tests I think that OS should be working normally and tests should take into consideration possible influence of the OS like systick handling, task scheduling, ticker interrupts, etc. So I suggest to use second bullet for upper layer tests.

Reply:
Since one of the problems is:
>The Timeout object used by the lp ticker wrapper locks deep sleep at various times causing tests which assert that deep sleep is allowed to fail

we could modify the tests to wait for the deep sleep to be unlocked before calling `sleep()`. But even this would require using critical section, so would also introduce more complexity to tests.

Adding `sleep_manager_can_deep_sleep_test()` seems straight forward, but such change would mean we accept that tests are occasionally run in an environment different from what they were supposed to. Simply put, the test result can be meaningless sometimes.

In my opinion the most honest solution would be adding
```cpp
#if DEVICE_LPTICKER && (LPTICKER_DELAY_TICKS > 0)
#error [NOT_SUPPORTED] Test not supported for this target
#endif
```
or executing affected test cases conditionally based on `LPTICKER_DELAY_TICKS > 0`.